From 853715b283b264471b39365e18fc2f3a273bc2a7 Mon Sep 17 00:00:00 2001
From: Matthew Lugg <mlugg@mlugg.co.uk>
Date: Sat, 29 Nov 2025 11:55:36 +0000
Subject: [PATCH 069/133] Sema: initialize OPV comptime allocs correctly
X-Developer-Signature: v=1; a=openpgp-sha256; l=1871; i=xtex@astrafall.org;
 h=from:subject; bh=oiWS83SsH3f9QzFFfH6BJINa8dlTRNYV2CK6YylkQoA=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78lh+zL1mNufk/XfLXt47o9jPo3lAx8f5qlDiliPax
 xwWhoRt6JjIwiDGxWAppshSZNjgzaqTzi+6rFwWZg4rE8gQaZEGBgYGBhYGvtzEvFIjHSM9U21D
 PUNDHSMdIwYuTgGY6tt8jAy3SrbI3zx2/IGuzV8/3WOLwhrn6uze7cV0Nup2Tdb8v4xmDP9L5od
 9iJ5x/VxYz66tLMbmZQZ9dQY/2U6s3MP4pjnirBgbAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

This was caused a `[0]std.builtin.Type.StructField.Attributes` to be
considered `undefined`, even though that type is OPV so should prefer
its OPV `.{}` over `undefined`.

Resolves: #30039
(cherry picked from commit 8f5db19791dde73e8e969c6ebf821767f1c3e50b)
---
 src/Sema.zig           | 4 +++-
 test/behavior/type.zig | 9 +++++++++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/Sema.zig b/src/Sema.zig
index 2dcf6be98fc7..dfb4647dff51 100644
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -175,9 +175,11 @@ const ComptimeAlloc = struct {
 
 /// `src` may be `null` if `is_const` will be set.
 fn newComptimeAlloc(sema: *Sema, block: *Block, src: LazySrcLoc, ty: Type, alignment: Alignment) !ComptimeAllocIndex {
+    const pt = sema.pt;
+    const init_val = try sema.typeHasOnePossibleValue(ty) orelse try pt.undefValue(ty);
     const idx = sema.comptime_allocs.items.len;
     try sema.comptime_allocs.append(sema.gpa, .{
-        .val = .{ .interned = try sema.pt.intern(.{ .undef = ty.toIntern() }) },
+        .val = .{ .interned = init_val.toIntern() },
         .is_const = false,
         .src = src,
         .alignment = alignment,
diff --git a/test/behavior/type.zig b/test/behavior/type.zig
index 78295d3a7de2..9d65e8e4350d 100644
--- a/test/behavior/type.zig
+++ b/test/behavior/type.zig
@@ -812,3 +812,12 @@ test "undefined type value" {
     };
     comptime assert(@TypeOf(S.undef_type) == type);
 }
+
+test "reify struct with zero fields through const arrays" {
+    const names: [0][]const u8 = .{};
+    const types: [0]type = .{};
+    const attrs: [0]std.builtin.Type.StructField.Attributes = .{};
+    const S = @Struct(.auto, null, &names, &types, &attrs);
+    comptime assert(@typeInfo(S) == .@"struct");
+    comptime assert(@typeInfo(S).@"struct".fields.len == 0);
+}
-- 
2.52.0

